### PR TITLE
A11y/Menu: Add keyboard support to Menu component

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Adds basic keyboard support to the Menu component.

See: #35613

### CODE REVIEW

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx [hunk 1]
```diff
@@ -53,37 +57,40 @@ export const MenuItem: React.FC<MenuItemProps> = React.memo(
                 }
               : undefined
           }
-          rel={target === '_blank' ? 'noopener noreferrer' : undefined}
+          role={url === undefined ? 'menuitem' : undefined}
+          data-role="menuitem" // used to identify menuitem in Menu.tsx
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:61]: Think it's worth making a note about what's going on here.
In general, the "menuitem" role is used to indicate an item within a menu that executes some kind of action upon being selected, like an item in a desktop application's toolbar. Since the MenuItem component can be a link, however, the "menuitem" role makes less sense, and assigning it to a link obscures the fact that clicking on it would navigate the browser to a different page. In light of this, I've chosen to only apply the "menuitem" role when the MenuItem is not a link.
Since we can't rely then on the `role` attribute to identify MenuItems, I've added the "data-role" attribute for that purpose.

File: packages/grafana-ui/src/components/Menu/Menu.tsx [hunk 2]
```diff
@@ -1,23 +1,97 @@
...
+      onOpen?.(setFocusedItem);
+    });
+
+    const handleKeys = (event: React.KeyboardEvent) => {
+      const menuItemsCount = wrapperRef?.current?.querySelectorAll('[data-role="menuitem"]').length ?? 0;
+
+      switch (event.key) {
+        case 'ArrowUp':
+          event.preventDefault();
+          event.stopPropagation();
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Menu/Menu.tsx:53]: These can be called before the switch, so they're not repeated for each case.

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx [hunk 3]
```diff
@@ -23,26 +23,30 @@ export interface MenuItemProps<T = any> {
...
+  React.forwardRef<HTMLAnchorElement & HTMLButtonElement, MenuItemProps>(
+    ({ url, icon, label, ariaLabel, target, onClick, className, active, tabIndex = -1 }, ref) => {
+      const styles = useStyles2(getStyles);
+      const itemStyle = cx(
+        {
+          [styles.item]: true,
+          [styles.activeItem]: active,
+        },
+        className
+      );
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:41]:
```suggestion
      const itemStyle = cx(
        {
          [styles.activeItem]: active,
        },
        styles.item,
        className
      );
```

File: packages/grafana-ui/src/components/ContextMenu/ContextMenu.story.tsx [hunk 4]
```diff
@@ -24,14 +24,17 @@ const menuItems = [
...
+      { label: 'Fifth', ariaLabel: 'Fifth' },
     ],
   },
 ];
 
 const renderMenuItems = () => {
   return menuItems?.map((group, index) => (
     <MenuGroup key={`${group.label}${index}`} label={group.label} ariaLabel={group.label}>
-      {(group.items || []).map((item) => (
+      {(group.items ?? []).map((item) => (
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/ContextMenu/ContextMenu.story.tsx:37]: Wonder why this change? In this case if `group.items` somehow ends being an empty string or `0` the code will throw an error. Also I think `group.items?.map` should work here.

File: packages/grafana-ui/src/components/Menu/Menu.tsx (hunk 2 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Menu/Menu.tsx:53]: Would this not eat certain key-combinations that we want to propagate (e.g. global shortcuts)?

Comment (Clarity-89) [reply packages/grafana-ui/src/components/Menu/Menu.tsx:53]: True, didn't think about that, good point.

File: packages/grafana-ui/src/components/ContextMenu/ContextMenu.story.tsx (hunk 4 above)
Comment (ashharrison90) [reply packages/grafana-ui/src/components/ContextMenu/ContextMenu.story.tsx:37]: +1 for `group.items?.map`, although that will still have the same problem if `group.items` is somehow `''` or `0`. i think that's extremely unlikely though. typescript should be stopping that from happening.

File: packages/grafana-ui/src/components/ContextMenu/ContextMenu.tsx [hunk 5]
```diff
@@ -41,12 +41,20 @@ export const ContextMenu: React.FC<ContextMenuProps> = React.memo(
     }, [x, y]);
 
     useClickAway(menuRef, () => {
-      if (onClose) {
-        onClose();
-      }
+      onClose?.();
```
Comment (ashharrison90) [inline packages/grafana-ui/src/components/ContextMenu/ContextMenu.tsx:44]: this is neat 🤓

File: packages/grafana-ui/src/components/Menu/Menu.tsx [hunk 6]
```diff
@@ -1,23 +1,97 @@
...
+      });
+    }, [localRef, focusedItem]);
+
+    useEffectOnce(() => {
+      const firstMenuItem = localRef?.current?.querySelector(`[data-role="menuitem"]`) as MenuItemElement | null;
+      if (firstMenuItem) {
+        firstMenuItem.tabIndex = 0;
+      }
+      onOpen?.(setFocusedItem);
+    });
```
Comment (ashharrison90) [inline packages/grafana-ui/src/components/Menu/Menu.tsx:45]: is there an advantage to using this over just using `useEffect` with an empty dependency array?

**Review by ashharrison90 — COMMENTED** [review]
tiny question

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx [hunk 7]
```diff
@@ -23,26 +23,30 @@ export interface MenuItemProps<T = any> {
...
 
-    return (
-      <div className={itemStyle} aria-label={ariaLabel}>
-        <a
-          href={url ? url : undefined}
+      const Wrapper = url === undefined ? 'button' : 'a';
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:43]: Maybe it would be easier to split this in 2 components?

**Review by hugohaggmark — APPROVED** [review]
Looks great and works great!

File: packages/grafana-ui/src/components/Menu/Menu.tsx (hunk 6 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Menu/Menu.tsx:45]: Just that our linter complains if you have an empty dependency array (which is a bit annoying tbh)

File: packages/grafana-ui/src/components/Menu/Menu.tsx [hunk 8]
```diff
@@ -1,23 +1,97 @@
...
+      }
+    };
+
     return (
-      <div {...otherProps} ref={ref} className={styles.wrapper} aria-label={ariaLabel}>
+      <div
+        {...otherProps}
+        ref={localRef}
+        className={styles.wrapper}
+        role="menu"
```
Comment (thisisobate) [inline packages/grafana-ui/src/components/Menu/Menu.tsx:90]: I think this should be able to serve two purposes: menu and listbox; case in point is the `refreshPicker` component. wdyt?

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx [hunk 9]
```diff
@@ -53,37 +57,40 @@ export const MenuItem: React.FC<MenuItemProps> = React.memo(
                 }
               : undefined
           }
-          rel={target === '_blank' ? 'noopener noreferrer' : undefined}
+          role={url === undefined ? 'menuitem' : undefined}
```
Comment (thisisobate) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:60]: same applies here...since parent container can either be menu or listbox, child list should have role menuitem or option.

File: packages/grafana-ui/src/components/Menu/Menu.tsx (hunk 8 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Menu/Menu.tsx:90]: I'm not sure `listbox` would be the appropriate role for the refresh picker, since it's a button that secretes a menu from which an option is selected. I will say though that for that particular component, the items should have the role "menuitemradio" since it indicates a persisted "choice" rather than an action that will be executed.
